sbi/neural_nets/density_estimators/flow.py

```diff
@@ -0,0 +1,60 @@
+from typing import Optional, Union
+from warnings import warn
+
+import torch
+from pyknos.nflows import flows
+from torch import Tensor, nn
+
+from sbi.neural_nets.density_estimators.base import DensityEstimator
+
+
+class NFlowsNSF(DensityEstimator):
+    r"""`nflows`- based neural spline flow density estimator.
+    Flow type objects already have a .log_prob() and .sample() method, so here we just wrap them and add the .loss() method.
+    """
+
+    def __init__(
+        self,
+        net: flows.Flow,
+    ):
+
+        super().__init__(net)
+
+    def log_prob(self, input: Tensor, condition: Tensor) -> Tensor:
+        r"""Return the batched log probabilities of the inputs given the conditions.
+
+        Args:
+            input: Inputs to evaluate the log probability of. Must have a batch dimension.
+            condition: Conditions. Must have a batch dimension.
+
+        Returns:
+            Sample-wise log probabilities.
+        """
+        return self.net.log_prob(input, context=condition)
+
+    def loss(self, input: Tensor, condition: Tensor) -> Tensor:
+        r"""Return the loss for training the density estimator.
+
+        Args:
+            input: Inputs to evaluate the loss on. Must be batched.
+            condition: Conditions. Must be batched.
+
+        Returns:
+            Average negative log-probability.
+        """
+        # We return the mean of the batch.
+        return -self.log_prob(input, condition).mean()
```

Comment:
This should not mean across batch-elements. Please also write a test for this.

Reply:
I believe the final resolution was to not take the mean at all and just return the per-sample loss (e.g. if we want to use calibration kernels in SNPE)